[NCCL] Tests for WorkNCCL::wait with Timeouts

osalpekar opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #41265 [Gloo] Tests for Gloo Async Work Wait-level Timeouts
* #40948 [Gloo] Support work-level timeouts in ProcessGroupGloo
* **#40947 [NCCL] Tests for WorkNCCL::wait with Timeouts**
* #40946 [NCCL] Support Wait Timeout in ProcessGroupNCCL
* #40944 [NCCL] Add timeout to ProcessGroup Work Wait

This PR adds tests for work-level timeouts in WorkNCCL objects. We kick off an allgather operation that waits for 1000ms before actually starting computation. We wait on completion of this allgather op with a timeout of 250ms, expecting the operation to timeout and throw a runtime error.

Differential Revision: [D22173101](https://our.internmc.facebook.com/intern/diff/D22173101/)

osalpekar committed:
[NCCL] Tests for WorkNCCL::wait with Timeouts

This PR adds tests for work-level timeouts in WorkNCCL objects. We kick off an allgather operation that waits for 1000ms before actually starting computation. We wait on completion of this allgather op with a timeout of 250ms, expecting the operation to timeout and throw a runtime error.

Differential Revision: [D22173101](https://our.internmc.facebook.com/intern/diff/D22173101/)

[ghstack-poisoned]

osalpekar committed:
Update on "[NCCL] Tests for WorkNCCL::wait with Timeouts"


This PR adds tests for work-level timeouts in WorkNCCL objects. We kick off an allgather operation that waits for 1000ms before actually starting computation. We wait on completion of this allgather op with a timeout of 250ms, expecting the operation to timeout and throw a runtime error.

Differential Revision: [D22173101](https://our.internmc.facebook.com/intern/diff/D22173101/)

[ghstack-poisoned]

osalpekar committed:
Update on "[NCCL] Tests for WorkNCCL::wait with Timeouts"


This PR adds tests for work-level timeouts in WorkNCCL objects. We kick off an allgather operation that waits for 1000ms before actually starting computation. We wait on completion of this allgather op with a timeout of 250ms, expecting the operation to timeout and throw a runtime error.

Differential Revision: [D22173101](https://our.internmc.facebook.com/intern/diff/D22173101/)

[ghstack-poisoned]

mrshenli commented on torch/lib/c10d/test/ProcessGroupNCCLTest.cpp:
```diff
@@ -386,6 +403,45 @@ void testAllgather(const std::string& path, int rank, int size) {
+    // First we reset the NCCL_BLOCKING_WAIT environment variable to prevent
+    // any unwanted side-effects. We must do this at all exit-points for this
+    // function.
+    setenv(c10d::NCCL_BLOCKING_WAIT, originalBlockingWait, 1);
```
nit (please feel free to ignore): does it worth it to make this a guard?

jiayisuse replied:
Good point. It's kind of overhead to restore original env value

mrshenli commented on torch/lib/c10d/test/ProcessGroupNCCLTest.cpp:
```diff
@@ -386,6 +403,45 @@ void testAllgather(const std::string& path, int rank, int size) {
+  setenv(c10d::NCCL_BLOCKING_WAIT, originalBlockingWait, 1);
+  // If no exception, that is also an error since we expect the test.wait call
+  // to timeout and throw.
+  throw std::runtime_error("BOOM!");
```
Can we make the message more informative here? Same for the above two throws.

osalpekar replied:
I was actually planning on cleaning up these messages, but just followed the same error messages as the already existing tests for now (they all do `throw std::runtime_error("BOOM!")` for some reason...). I'm planning on putting up another PR that fixes all of these.

osalpekar committed:
Update on "[NCCL] Tests for WorkNCCL::wait with Timeouts"


This PR adds tests for work-level timeouts in WorkNCCL objects. We kick off an allgather operation that waits for 1000ms before actually starting computation. We wait on completion of this allgather op with a timeout of 250ms, expecting the operation to timeout and throw a runtime error.

Differential Revision: [D22173101](https://our.internmc.facebook.com/intern/diff/D22173101/)

[ghstack-poisoned]

osalpekar committed:
Update on "[NCCL] Tests for WorkNCCL::wait with Timeouts"


This PR adds tests for work-level timeouts in WorkNCCL objects. We kick off an allgather operation that waits for 1000ms before actually starting computation. We wait on completion of this allgather op with a timeout of 250ms, expecting the operation to timeout and throw a runtime error.

Differential Revision: [D22173101](https://our.internmc.facebook.com/intern/diff/D22173101/)

[ghstack-poisoned]

osalpekar committed:
Update on "[NCCL] Tests for WorkNCCL::wait with Timeouts"


This PR adds tests for work-level timeouts in WorkNCCL objects. We kick off an allgather operation that waits for 1000ms before actually starting computation. We wait on completion of this allgather op with a timeout of 250ms, expecting the operation to timeout and throw a runtime error.

Differential Revision: [D22173101](https://our.internmc.facebook.com/intern/diff/D22173101/)

[ghstack-poisoned]

osalpekar committed:
Update on "[NCCL] Tests for WorkNCCL::wait with Timeouts"


This PR adds tests for work-level timeouts in WorkNCCL objects. We kick off an allgather operation that waits for 1000ms before actually starting computation. We wait on completion of this allgather op with a timeout of 250ms, expecting the operation to timeout and throw a runtime error.

Differential Revision: [D22173101](https://our.internmc.facebook.com/intern/diff/D22173101/)

[ghstack-poisoned]
